### PR TITLE
 #2198 Refactor checker to SettingDef

### PR DESCRIPTION
**Pull request by saivirtue**
### Checklist
- [x] [Review Contribution Guidelines](https://github.com/oharastream/ohara/blob/master/CONTRIBUTING.md)
- [x] Add Reviewers (see [Authors](https://github.com/oharastream/ohara#authors))
- [x] Add Assignees (of course, you are the assignee by default!)
- [x] DON'T set a Milestone
- [x] DON'T add any labels (PR labels will be added automatically by Jenkins)
- [x] Link to an issue(You can do so by using the required by keyword: required by #2198)
- [x] Add unit tests (or please explain why this PR is not worth having new tests)
- [x] Pass QA (Sometimes you may encounter an existing flaky test which would obstruct you from getting QA passed. If the failed tests are unrelated, you can add a comment in the PR thread)

### CODE REVIEW

File: ohara-common/src/main/java/com/island/ohara/common/setting/SettingDef.java [hunk 1]
```diff
@@ -163,90 +168,160 @@ private SettingDef(
...
-            } catch (Exception e) {
+    return (Object value) -> {
+      // we don't check the optional key with null value
+      if (!this.required && this.defaultValue == null && value == null) return;
+
+      // besides the first rule, any other combination of settings should check the value is not
+      // null (default or from api)
+      Object trueValue = value == null ? this.defaultValue : value;
+      if (trueValue == null)
+        throw new OharaConfigException("the key [" + this.key + "] value is required");
```
Comment (chia7712) [inline ohara-common/src/main/java/com/island/ohara/common/setting/SettingDef.java]: remove the word "value"

File: ohara-common/src/main/java/com/island/ohara/common/exception/OharaConfigException.java [hunk 2]
```diff
@@ -39,6 +39,6 @@ public OharaConfigException(String name, Object value, String message) {
     super(
         String.format(
             "Invalid value of [%s] for configuration [%s] :\n%s",
-            name, value, message == null ? "" : message));
+            value, name, message == null ? "" : message));
```
Comment (chia7712) [inline ohara-common/src/main/java/com/island/ohara/common/exception/OharaConfigException.java]: please add enough comment for OharaConfigException class. Now, it is important exception in our checker.

Comment (chia7712) [inline ohara-common/src/main/java/com/island/ohara/common/exception/OharaConfigException.java]: please highlight the type also.

**Comment by saivirtue** [comment]
> a nit. seems this PR aims to "refactor" than "add" checker. Could you revise the title?

Sure.

File: ohara-common/src/main/java/com/island/ohara/common/setting/SettingDef.java (hunk 1 above)
Comment (saivirtue) [reply ohara-common/src/main/java/com/island/ohara/common/setting/SettingDef.java]: my bad :(

File: ohara-common/src/main/java/com/island/ohara/common/exception/OharaConfigException.java (hunk 2 above)
Comment (saivirtue) [reply ohara-common/src/main/java/com/island/ohara/common/exception/OharaConfigException.java]: ok

**Comment by saivirtue** [comment]
retry client

File: ohara-common/src/main/java/com/island/ohara/common/setting/SettingDef.java [hunk 3]
```diff
@@ -163,90 +168,162 @@ private SettingDef(
...
+          try {
+            // try parse the json string to Connector Key
+            ConnectorKey.toConnectorKey(String.valueOf(trueValue));
+          } catch (Exception e) {
+            throw new OharaConfigException(this.key, trueValue, e.getMessage());
+          }
+          break;
+        case TAGS:
+          try {
+            parseTags(trueValue);
```
Comment (chia7712) [inline ohara-common/src/main/java/com/island/ohara/common/setting/SettingDef.java]: why move the impl to be a method? Is it used by other classes?

Comment (saivirtue) [reply ohara-common/src/main/java/com/island/ohara/common/setting/SettingDef.java]: It's the habit - I prefer to separate the actual implementation to other place for better review.
It is ok to move the impl back

**Comment by saivirtue** [comment]
retry it

**Comment by saivirtue** [comment]
retry it

**Comment by saivirtue** [comment]
The failed tests are not releated, merge now.

**Pull request merged**